File: Python/RedundantConnection.py
## Leetcode Problem - 684

class UnionFind:
    def __init__(self,n) -> None:
        self.par = {}
        self.rank = {}

        for i in range(1,n+1):
            self.par[i] = i
            self.rank[i] = 0

    def find(self,n):
        p = self.par[n]

        while p!= self.par[p]:
            self.par[p] = self.par[self.par[p]]
            p = self.par[p]
        
        return p

    def union(self,n1,n2):
        p1,p2 = self.find(n1),self.find(n2)
        
        if p1 == p2:
            return False
        
        if self.rank[p1] > self.rank[p2]:
            self.par[p2] = p1
        elif self.rank[p2] > self.rank[p1]:
            self.par[p1] = p2
        else:
            self.par[p1] = p2
            self.rank[p2] += 1

        return True



def findRedundantConnection(edges):

    n = 0
    for u,v in edges:
        n = max(u,max(v,n))

    unionFind = UnionFind(n)

    result = []

    for u,v in edges:
        if not unionFind.union(u,v):
            result = [u,v] 

    return result
    


edges = [[1,2],[1,3],[2,3]]
print(findRedundantConnection(edges))

File: Javascript/RedundantConnection.js
// Leetcode Problem - 684

var UnionFind = function(n){
    this.par = new Map()
    this.rank = new Map()

    for (let i=1;i<n+1;i++){
        this.par.set(i,i)
        this.rank.set(i,0)
    }
}


UnionFind.prototype.find = function(n){
    let p = this.par.get(n)

    while (p!==this.par.get(p)){
        this.par.set(p,this.par.get(this.par.get(p)))
        p = this.par.get(p)
    }

    return p
}

UnionFind.prototype.union = function(n1,n2){
    let p1 = this.find(n1)
    let p2 = this.find(n2)

    
    if (p1===p2){
        return false
    }

    if (this.rank.get(p1)>this.rank.get(p2)){
        this.par.set(p2,p1)
    }
    else if (this.rank.get(p2)>this.rank.get(p1)){
        this.par.set(p1,p2)
    }
    else{
        this.par.set(p1,p2)
        this.rank.set(p2,this.rank.get(p2)+1)
    }

    return true
}

var findRedundantConnection = function(edges){

    let n = 0
    for (const [u,v] of edges){
        n = Math.max(Math.max(v,n),u)
    }

    let unionfind = new UnionFind(n)

    result = []
    for (const [u,v] of edges){
        if (!unionfind.union(u,v)){
            result = [u,v]
        }
    }

    return result
    
}


edges = [[1,2],[1,3],[2,3]]
console.log(findRedundantConnection(edges))